master/internal/provisioner/aws_spot_legacy.go

```diff
@@ -0,0 +1,216 @@
+package provisioner
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+
+	"github.com/determined-ai/determined/master/pkg/actor"
+)
+
+// There was a bug where spot requests and instances were not being tagged with
+// the resource pool, leading to multiple resource pools trying to manage the
+// same instances. The code now uses a resource pool tag, but that means that
+// instances with the old tag format are ignored. To make sure we aren't leaving
+// orphaned instances after a version upgrade, we identify instances with the old
+// format and clean them up. Because the spot API is eventually consistent, we
+// repeat the cleanup after 5 minutes. This function can be removed once all users
+// are on versions newer than 0.15.5.
+func (c *awsCluster) cleanupLegacySpotInstances(ctx *actor.Context) {
+	loggerSpotLegacy := ctx.Log().WithField("codepath", "spotLegacy")
+
+	// Repeat after 5 minutes to handle the fact that the AWS API is eventually consistent.
+	go func() {
+		loggerSpotLegacy.Infof("starting legacy spot cleanup operation")
+		c.legacyCleanupActiveSpotRequestsAndInstances(ctx)
+		c.legacyCleanupCanceledButInstanceRunningSpot(ctx)
+
+		time.Sleep(5 * time.Minute)
+
+		loggerSpotLegacy.Debugf("starting second pass of legacy spot cleanup")
+		c.legacyCleanupActiveSpotRequestsAndInstances(ctx)
+		c.legacyCleanupCanceledButInstanceRunningSpot(ctx)
+		loggerSpotLegacy.Debugf("completed legacy spot cleanup")
+	}()
+}
+
+func (c *awsCluster) legacyCleanupActiveSpotRequestsAndInstances(ctx *actor.Context) {
+	loggerSpotLegacy := ctx.Log().WithField("codepath", "spotLegacy")
+
+	// List spot requests with the old format
+	activeSpotReqs, err := c.legacyListActiveSpotInstanceRequests(ctx)
+	if err != nil {
+		loggerSpotLegacy.
+			WithError(err).
+			Errorf("cannot list active legacy spot requests")
+		return
+	}
+	if activeSpotReqs.numReqs() == 0 {
+		loggerSpotLegacy.Debugf("no active legacy spot requests to clean up")
+		return
+	}
+	// Delete spot requests with old format
+	loggerSpotLegacy.Infof(
+		"terminating %d active legacy spot requests: %s",
+		activeSpotReqs.numReqs(),
+		strings.Join(activeSpotReqs.idsAsList(), ","))
+
+	_, err = c.terminateSpotInstanceRequests(ctx, activeSpotReqs.idsAsListOfPointers(), false)
+	if err != nil {
+		loggerSpotLegacy.
+			WithError(err).
+			Errorf("unable to terminate active legacy spot requests")
+	}
+
+	// Delete spot instances associated with the active requests
+	instancesToTerminate := newSetOfStrings()
+	for _, req := range activeSpotReqs.iter() {
+		if req.InstanceID != nil {
+			instancesToTerminate.add(*req.InstanceID)
+		}
+	}
```

Comment:
why not `instancesToTerminate := activeSpotReqs.instanceIds()`?

Reply:
fixed